<!doctype html>
<html lang='en'>
  <head>
    <meta charset='utf-8'>

    <title>Automatic Revision of Ecological Theories</title>

    <meta name='description' content='Thesis defense'>
    <meta name='author' content='Philippe Desjardins-Proulx'>

    <meta name='apple-mobile-web-app-capable' content='yes' />
    <meta name='apple-mobile-web-app-status-bar-style' content='black-translucent' />

    <meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no'>

    <link rel='stylesheet' href='css/reveal.min.css'>
    <link rel='stylesheet' href='css/theme/thesis.css' id='theme'>
    <link rel='stylesheet' href='lib/css/zenburn.css'>

    <script>
      if (window.location.search.match(/print-pdf/gi)) {
        var link = document.createElement('link');
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = 'css/print/pdf.css';
        document.getElementsByTagName('head')[0].appendChild(link);
      }
    </script>

    <!--[if lt IE 9]>
    <script src='lib/js/html5shiv.js'></script>
    <![endif]-->
  </head>

  <body>

    <div class='reveal'>
      <!-- Any section element inside of this container is displayed as a slide -->
      <div class='slides'>
        <section>
          <section>
            <h1>Automatic Revision<br/> of Ecological Theories</h1>
            <p>by <a href='http://phdp.github.io/'>Philippe Desjardins-Proulx</a></small></p>
          </section>
        </section>
        <section>
          <section>
            <h1>Motivation</h1>
          </section>
          <section>
            <img src='img/amnat12.png' alt='amnat12'/>
          </section>
          <section>
            <img src='img/eco-evo-paper.png' alt='ecoevo'/>
          </section>
          <section>
            <h2>Species Richness/Speciation Relationship</h2>
            <img width='60%' height='60%' src='img/complex-1.svg' alt='res'/>
            <p>Species richness (green) and speciation events (blue) on simulations with 5 (top) and 25 (bottom) communities.</p>
          </section>
          <section>
            <h2>The Big Picture</h2>
            <img src='img/gillespie1997.png' alt='gil04'/>
            <p><i>Reference: JH Gillespie. Population Genetics: A Concise Guide, 1st edition. Johns Hopkins University Press, 1997.</i></p>
          </section>
          <section>
            <img src='img/hahn08.png' alt='hahn08'/>
          </section>
          <section>
            <h2>Why Biodiversity is so Complicated?</h2>
            <p>The effect of mutations vary greatly across the genome.</p>
            <p>Selection varies in time/space as the biotic/abiotic environment changes.</p>
            <p>The influence of linkage and recombination.</p>
            <p>Speciation has a complex spatial structure.</p>
          </section>
          <section>
            <h3>What if there is no<br/>single theory of biodiversity?</h3>
          </section>
          <section>
            <h3>What if the problem is to reason<br/>about a <i>set</i> of ecological theories?</h3>
          </section>
          <section>
            <h2>Two Key Questions</h2>
            <br/>
            <h3>How can we formally represent ecological theories<br/>(a different question that how to think about data)?</h3>
            <br/>
            <h3>How can we exploit this formal representation<br/>(inference, discovery and revision of theories)?</h3>
            <br/>
            <p>In practice, these questions lead to knowledge representation and machine learning.</p>
          </section>
          <section>
            <h2>Knowledge Representations</h2>
            <img src='img/reasoning_systems.svg' alt='reasoning'/>
          </section>
          <section>
            <h2>Machine Learning</h2>
            <img src='img/xkcd_machine_learning.png' alt='xkcd_ml'/>
            <p>Two main types: supervised learning, unsupervised learning.</p>
            <p><i>Reference: KP Murphy. Machine Learning: a Probabilistic Perspective. MIT Press, 2012.</i></p>
          </section>
          <section>
            <h2>Supervised Learning</h2>
            <img src='img/supervised.png' alt='supervised'/>
            <p>Given data points \(\mathcal{D} = \{(\mathbf{x}_i, y_i)\}_{i = 0}^{|\mathcal{D}| - 1}\), learn to predict \(y\) from \(\mathbf{x}\).</p>
            <p>Tasks: <i>classification, regression.</i></p>
          </section>
          <section>
            <h2>Unsupervised Learning</h2>
            <img src='img/matrix-completion.png' alt='matrix ompletion'/>
            <img src='img/KMeans-Gaussian-data.svg' alt='kmeans'/>
            <p>Given data points \(\mathcal{D} = \{(\mathbf{x}_i)\}_{i = 0}^{|\mathcal{D}| - 1}\), learn some 'hidden structure'.</p>
            <p>Tasks: <i>matrix completion, clustering, structure learning.</i></p>
          </section>
          <section>
            <h2>A Knowledge Base<br/>for Ecological Knowledge</h2>
            <p><b>Domains</b>: species \(s, s_i, s_j \in S\), locations \(l \in L\), time \(t \in \mathbb{R}\).</p>
            <p><b>Symbols</b>: universal quantifier (<i>forall</i>) \(\forall\), negation \(\neg\).</p>
            <br/>
            \[\forall s_i, s_j: PhyloDistance(s_i, s_j) > 0, \infty.\]
            \[\forall s_i, s_j: \dot{x} = \alpha s_i - \beta s_is_j \mbox{ and } \dot{s_j} = \delta s_is_j - \gamma s_j, 0.01.\]
            <p>...</p>
            \[\forall s, l, t: SmallN(s, l) \Rightarrow N(s,l,t+1) = G(s) \times N(s,l,t), 0.14.\]
            \[\forall s_i, s_j, l: KNN_5(s_i, s_j) > 0.5 \Rightarrow PreyOnAt(p, q, l), 2.5.\]
          </section>
          <section>
            <h2>Ecological Interactions as a<br/>Machine Learning Problem</h2>
            <p>Predict whether \(x\) interacts with \(y\) given traits for both species.</p>
            <p>Fill holes in an interaction matrix (matrix completion).</p>
            <p>Discover or revise logic and mathematical rules about interactions given data.</p>

            <br/>
            <p><i>Interactions assumed to involve only two species. A non-interaction is often implied by lack of observed interaction.</i></p>
          </section>
          <section>
            <h2>Why Ecological Interactions?</h2>
            <p>A key reason why selection is difficult to handle
            with traditional theories is because of the complexity of species
            interactions.</p>
            <br/>
            <p>My lab is studying interactions. I had access to
            both good data-sets and people who understood them well.</p>
          </section>
          <section>
            <h2>Machine Learning and Interactions Roadmap</h2>
            <img src='img/roadmap.svg.png' alt='roadmap'/>
          </section>
        </section>
        <section>
          <section>
            <h1>Classical Machine<br/>Learning</h1>
          </section>
          <section>
            <h2>Data</h2>
            <ul width='100%'>
              <li>Assembled by Isabelle Daigle, mostly from Digel et al. (2014).</li>
              <li>881 species.</li>
              <li>34 193 unique interactions.</li>
              <li>3 real valued-traits: \(body mass\), we also performed principal coordinates analysis on taxonomic distances, and use the scores of each species on the first two axes \(Ph_0\), \(Ph_1\).</li>
              <li>25 binary traits (e.g. AboveGround, Detritivore, Bacteria, Jumps, LongLegs, UsePoison).</li>
            </ul>
          </section>
          <section>
            <h2>Supervised Learning &amp; Recommender</h2>
            <img src='img/ecological-netflix-paper.png' alt='paper'/>
          </section>
          <section>
            <h2>True Skill Statistic</h2>
            \[TSS = \frac{(tp \times tn) - (fp \times fn)}{(tp + fn)(fp + tn)}.\]
            <br/><br/>
            <p>With:</p>
            <ul>
              <li><i>tp</i>: true positives</li>
              <li><i>tn</i>: true negatives</li>
              <li><i>fp</i>: false positives</li>
              <li><i>fn</i>: false negatives</li>
            </ul>
            <br/><br/>
            <p>The <i>TSS</i> ranges from -1 to 1.</p>
          </section>
          <section>
            <h2>Supervised Learning</h2>
            <img src='img/rf.png' alt='rf'/>
            <p>Using the binary and real-valued traits of both species as features.</p>
            <p>Tested several methods: random forests, support vector machines, decision trees.</p>
          </section>
<!--          <section>
            <h2>Similarity Measure</h2>
            \[tanimoto(\mathbf{x}, \mathbf{y}) = \frac{\left\vert\mathbf{x} \cap \mathbf{y}\right\vert}{\left\vert\mathbf{x} \cup \mathbf{y}\right\vert}.\]

            <br/>
            <p>For \(\mathbf{x} = \{A, B, C\}\) and \(\mathbf{y} = \{A, C, D, E, F\}\):</p>
            <br/>

\begin{align}
  tanimoto(\mathbf{x}, \mathbf{y}) &= \frac{\left\vert\{A, B, C\} \cap \{A, C, D, E, F\}\right\vert}{\left\vert\{A, B, C\} \cup \{A, C, D, E, F\}\right\vert}\\
                                   &= \frac{\left\vert\{A, C\}\right\vert}{\left\vert\{A, B, C, D, E, F\}\right\vert} = \frac{2}{6}.
\end{align}
          </section> -->
          <section>
            <h2>Recommendation</h2>
            <img src='img/recommendation-table.png' alt='rectable'/>
          </section>
          <section>
            <h2>Results: Supervised Learning</h2>

            <p>Random forests predict correctly 99.55% of the non-interactions
            and 96.81% of the interactions, for a <i>TSS</i> of 0.96.</p>

            <br/>

            <p>Removing the binary traits has little effect on the model. With
            only \(body mass\), \(Ph_0\), \(Ph_1\) (the first two axes from a
            principal coordinates analysis on taxonomic distances), the
            <i>TSS</i> of the random forest is 0.94.</p>
          </section>
          <section>
            <h2>Results: Recommendation</h2>
            <img width='80%' src='img/recommendation.svg' alt='rec'/>
          </section>
          <section>
            <h2>Bartomeus' Pollinator-Plant Data-Set</h2>
            <ul width='100%'>
              <li>Plant-pollinator interactions were collected across 16 sites in
              the southwest of the iberian peninsula (Huelva and Seville).</li>
              <li>65 pollinators.</li>
              <li>277 plants.</li>
              <li>739 pollinator-plant interactions.</li>
              <li>We randomly pick 739 pollinator-plant pairs that are not
              interacting to build a data-set of 1478 entries, half of
              interactions, half of non-interactions.</li>
            </ul>
          </section>
          <section>
            <h2>Matrix Completion<br/>with the K-Nearest Neighbors (KNN)</h2>
            <img width='100%' src='img/knn-poll.svg' alt='knn-poll'/>
          </section>
          <section>
            <h2>KNN with Threshold</h2>
            <img src='img/full-confusion.png' alt='confusion'/>

            <p>Full confusion matrix for K = 3 and threshold of 0.0 and 0.5
            (majority vote). The accuracies are of 0.929 and 0.975. TSS scores
            are 0.827 vs 0.710.</p>
          </section>
          <section>
            <h2>Classical Machine Learning &amp; Interactions</h2>
            <ul>
              <li>We confirm the old result that body mass is key to how interactions are structured (Williams &amp; Martinez 2000).</li>
              <li>Phylogeny, combined with body mass, is an important feature.</li>
              <li>KNN-based algorithms exploit nestedness: how much species cluster around the same set of preys in a food web.</li>
              <li>Interactions are fairly easy to predict (TSS really close to 1).</li>
              <li>Binary traits contain little useful information.</li>
            </ul>
          </section>
        </section>
        <section>
          <section>
            <h1>Probabilistic Logic</h1>
          </section>
          <section>
            <h1>Markov Logic</h1>
            <p>Knowledge base \(\mathcal{M}:\)</p>
            \[\forall x, y, z: Friends(x, y) \mbox{ and } Friends(y, z) \Rightarrow Friends(x, z), 0.7.\]
            \[\forall x: Smoking(x) \Rightarrow Cancer(x), 1.5.\]
            \[\forall x, y: Friends(x, y) \mbox{ and } Smoking(x) \Rightarrow Smoking(y), 1.1.\]

            <br/>

            <p>Can answer queries of the form:</p>
            \[P(Cancer(Bob) | Smoking(Anna) \mbox{ and } Friends(Anna, Bob), \mathcal{M}, \mathcal{D}).\]
            \[P(Friends(x, y) \Leftrightarrow Friends(y, x) | \mathcal{M}, \mathcal{D}).\]

            <br/> <p>Predicates are bivalent (true or false) but formulas are
            weighted. The <b>cost</b> is the sum of the weights of the rules
            violated by a given truth assignment (world).</p>

            <!-- Weights as log odds between world probabilities. -->
          </section>
          <section>
            <h2>Predicate Logic and Food Webs</h2>
            <img src='img/metaweb.png' alt='metaweb'/>
            <p>Poisot et al. <i>The dissimilarity of species interaction networks</i>. Ecol Lett, 2012.</p>
            <br/>
            \[PreyOn(x, y), PreyOnAt(x, y, l), PreyOnInAbsence(x, y, z).\]
            <br/>
            \[\forall x, y: PreyOn(x, y) \mbox{ and } \neg Parasite(x) \Rightarrow M(x) > M(y).\]
          </section>
          <section>
            <h2>Salix Data-Set</h2>
            <img src='img/salix.png' alt='salix'/>
            \[Parasites \rightarrow Galler \rightarrow Salix\]
          </section>
          <section>
            <h2>Learning Weights and Rules with Alchemy2</h2>
<p style='color:#4A9E4E'>0: \(PreyOnAt(x,y,z) \Rightarrow PreyOn(x,y), 5.20.\)</p>
<p style='color:#4A9E4E'>1: \(HighCooccurrence(x,y) \Rightarrow PreyOn(x,y), 4.22.\)</p>
<p style='color:#4A9E4E'>2: \(IsGaller(x) \mbox{ and } PreyOn(x,y) \Rightarrow IsSalix(y), 4.15.\)</p>
<p style='color:#4A9E4E'>3: \(IsParasitoid(y) \mbox{ and } PreyOn(y,x) \Rightarrow IsGaller(x), 3.49.\)</p>
<p style='color:#4A9E4E'>4: \(PreyOn(x,y) \Rightarrow HighCooccurrence(x,y), 1.57.\)</p>
<p style='color:#BE1F1F'>5: \(PreyOn(x,y) \Rightarrow PreyOn(x,x), 1.52.\)</p>
<p style='color:#BE1F1F'>6: \(PreyOn(x,x) \Rightarrow PreyOn(x,y), 0.45.\)</p>
<p style='color:#BE1F1F'>7: \(\neg IsSalix(x) \mbox{ or } \neg PreyOn(x,y), 0.02.\)</p>
<p style='color:#4A9E4E'>8: \(CloselyRelated(x,y) \mbox{ and } PreyOn(x,z) \Rightarrow PreyOn(y,z), 0.00.\)</p>
          </section>
          <section>
            <h2>What Went Wrong?</h2>
            <p>Alchemy-2 has a slow (maybe inaccurate) L-BFGS implementation (i.e.: learning weights is slow).</p>
            <p>Lessons from classical methods: real-valued variables are important.</p>
          </section>
          <section>
            <h2>What Went Wrong?</h2>
            <img src='img/salix-b.png' alt='salix'/>
            <p style='color:#FF0000'>In red: predicates that would benefit from being many-valued.</p>
            <p style='color:#FFA700'>In orange: could be bivalent or many-valued.</p>
            <p style='color:#65D63E'>In green: bivalent predicates.</p>
          </section>
        </section>
        <section>
          <section>
            <h1>Fuzzy Logic</h1>
          </section>
          <section>
            <h2>Bartomeus' Pollinator-Plant Data-Set (Again)</h2>
            <ul width='100%'>
              <li>Plant-pollinator interactions were collected across 16 sites in the southwest of the iberian peninsula (Huelva and Seville).</li>
              <li>65 pollinators.</li>
              <li>277 plants.</li>
              <li>739 pollinator-plant interactions.</li>
              <li>We randomly pick 739 pollinator-plant pairs that are not
              interacting to build a data-set of 1478 entries, half of
              interactions, half of non-interactions.</li>
              <li style='color:#072DB0'>Seven real-valued traits: Plant's nectar tube dimension, Plant's nectar tube depth, Plant's flower width, Pollinator's body size, Pollinator's tongue length, Pollinator generalism, \(K\)-nearest neighbour ratio.</li>
            </ul>
          </section>
          <section>
            <h2>Structure</h2>
            \[\mathbf{If} \mbox{ } antecedants \mbox{ } \mathbf{then} \mbox { } consequents.\]
            \[\mathbf{If} \mbox{ } antecedants \mbox{ } \mathbf{then} \mbox { } consequents.\]
            \[\mathbf{If} \mbox{ } antecedants \mbox{ } \mathbf{then} \mbox { } consequents.\]
            \[...\]

            <br/>

            <p><b>If</b> Pollinator is specialist <b>and</b> Plant nectar tube dim is small <b>then</b> Interaction.</p>

          </section>
          <section>
            <h2>Linguistic Variables</h2>
            <img width='70%' height='60%' src='img/fsets.svg'/>
            <p>If the flower width is 6.5mm, then <b>"Flower width is High"</b> has a truth value of 0.75, <b>"Flower width is Very High"</b> has a truth alue of 0.25.</p>
          </section>
          <section>
            <h2>Initial Model</h2>
            <ul style='width: 100%'>
              <li style='color:#277F2A'><b>If</b> KNN is high <b>then</b> \(I^+\).</li>
              <li style='color:#7F2D27'><b>If</b> KNN is low <b>then</b> \(I^-\).</li>
            </ul>
            <br/><br/>
            <p>\(I^+\) denotes an interaction, \(I^-\) a non-interaction.</p>
          </section>
          <section>
            <h2>Learning Rules</h2>
            <ol start='0'>
              <li>Start with the two KNN rules.</li>
              <li>Randomly generate a rule, add it to the knowledge base.</li>
              <li>If the new knowledge base beats the old by \(\alpha\) TSS on the training set, keep the rule.</li>
              <li>Repeat #1-2 \(\phi\) times.</li>
            </ol>
            <br/>
            <br/>
            <p>Starts with high acceptance threshold \(\alpha\), lower it with time.</p>
            <br/>
            <p>C++14 code: <a href='https://github.com/PhDP/Manticore'>https://github.com/PhDP/Manticore</a></p>
          </section>
          <section>
            <h2>Results</h2>
            <ul style='width: 100%; font-size: 80%'>
              <li style='color:#277F2A'><b>If</b> Plant nectar tube depth is average <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> Pollinator body size is small <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> KNN is high <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> Pollinator is specialist <b>and</b> Plant nectar tube dim is small <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> Plant flower width is average <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> Plant flower width is small <b>and</b> Pollinator tongue length is small <b>then</b> \(I^+\).</li>
              <li style='color:#7F2D27'><b>If</b> KNN is low <b>then</b> \(I^-\).</li>
            </ul>
            <br/><br/>
            <p>\(I^+\) denotes an interaction, \(I^-\) a non-interaction.</p>
          </section>
          <section>
            <h2>Results</h2>
            <table style='width:100%; font-size:80%; text-align:center;'>
              <tr>
                <th>Model</th>
                <th>TSS (10% testing data)</th>
              </tr>
              <tr>
                <td>Fuzzy (initial model with two KNN rules)</td>
                <td>0.564</td>
              </tr>
              <tr>
                <td>Decision trees</td>
                <td>0.723</td>
              </tr>
              <tr>
                <td>Random forests</td>
                <td>0.795</td>
              </tr>
              <tr>
                <td>Fuzzy (after evolution)</td>
                <td>0.885</td>
              </tr>
            </table>
            <br/>
            <p>The <a href='http://scikit-learn.org/stable/index.html'>scikit-learn</a> Python package was used for decision trees and random forests. All algorithms used the TSS for training and testing.</p>
          </section>
          <section>
            <h2>Lessons</h2>
            <p>A few traits determine whether a pollinator will interact with a plant: nectar tube dimension, nectar tube depth, flower width, pollinator's body size, pollinator's tongue length.</p>
            <p>Possible to learn and revise easy-to-read ecological theories of interactions.</p>
            <p>The importance of real-valued variables for species interaction (again!).</p>
            <p>Should explore genetic algorithms for learning (e.g. NSGA-II).</p>
          </section>
        </section>
        <section>
          <section>
            <h1>Conclusion</h1>
          </section>
          <section>
            <h2>Two Key Questions</h2>
            <h3>How can we formally represent ecological theories?</h3>
            <h3>How can we exploit this formal representation?</h3>
            <img src='img/reasoning_systems.svg' alt='reasoning'/>
          </section>
          <section>
            <h2>In Brief</h2>
            <ul width='100%'>
              <li>Finding the proper knowledge representation for ecological theories remains an open question.</li>
              <li>Supervised learning algorithms (esp. random forests) can predict interactions with few <b>real-valued</b> traits.</li>
              <li>Fuzzy logic rules can predict pollinator-plant interactions with a few real-valued traits.</li>
              <li>A simple revision algorithm yields very good results for fuzzy logic.</li>
              <li>Markov logic has computational issues and is held back by its inability to handle nuanced predicates.</li>
            </ul>
          </section>
        </section>
        <section>
          <section>
            <h1>Thanks!</h1>
            <h3>Dominique Gravel, Timothée Poisot, Alireza Tamaddoni Nezhad, Shengrui Wang, Pierre-Étienne Jacques.</h3>
            <h3>Hedvig Nenzén, Guillaume Blanchet, Steve Vissault, Idaline Laigle, Ignasi Bartomeus.</h3>
          </section>
        </section>

      </div>
    </div>

    <script src='lib/js/head.min.js'></script>
    <script src='js/reveal.min.js'></script>
    <script>
      // Full list of configuration options available here:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        width: '70%',
        height: '100%',
        margin: 0,
        minScale: 1,
        maxScale: 1,
        controls: true,
        progress: true,
        history: true,
        center: true,
        slideNumber: 'c/t',

        theme: Reveal.getQueryHash().theme, // available themes are in /css/theme
        transition: Reveal.getQueryHash().transition || 'linear', // default/cube/page/concave/zoom/linear/fade/none

        // Parallax scrolling
        // parallaxBackgroundImage: 'https://s3.amazonaws.com/hakim-static/reveal-js/reveal-parallax-1.jpg',
        // parallaxBackgroundSize: '2100px 900px',

        // Optional libraries used to extend on reveal.js
        dependencies: [
          { src: 'lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
          { src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
          { src: 'plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
          { src: 'plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
          { src: 'plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
        ]
      });
    </script>
    <script type='text/javascript' src='../../js/bower/MathJax/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
